In [3]:
from matplotlib import rc
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import os
import pickle

import warnings
warnings.filterwarnings('ignore')


rc('font', family='AppleGothic')
plt.rcParams['axes.unicode_minus'] = False

# DataFrame 변수 List
- 경쟁사이용: competitor
- 고객DOMO: demo
- 구매상품 TR: purprd
- 멤머십여부: member
- 상품분류: prodcl
- 채널이용: channel
- 분류:
- 분류로직:

# Data List

In [5]:
file_list = os.listdir('../../data')
sorted(file_list)

['.DS_Store',
 '테이블명_컬럼명.png',
 '경쟁사이용.txt',
 '고객DEMO.txt',
 '구매상품TR.txt',
 '멤버십여부.txt',
 '분류.docx',
 '분류.txt',
 '분류로직.csv',
 '분류로직.xlsx',
 '상품분류.txt',
 '채널이용.txt']

## 1. 경쟁사이용

In [11]:
# 경쟁사이용
competitor = pd.read_csv('../../data/경쟁사이용.txt', encoding='cp949')
competitor.head()

고객번호 제휴사  경쟁사    이용년월
0     2   D  D02  201507
1    51   D  D01  201504
2    77   D  D02  201503
3    77   D  D02  201506
4    77   D  D02  201507

- competitor.고객번호.unique()
> len: 5881\
> 1 ~ 19377
- competitor.제휴사.unique()
> ['A', 'B', 'C', 'D']
- competitor.경쟁사.unique()
> ['A01', 'A02', 'B01', 'B02', 'C01', 'C02', 'C03', 'D01', 'D02']
- competitor.이용년월.unique()
> [201501, 201502, 201503, 201504, 201505, 201506, 201507, 201508, 201509, 201510, 201511, 201512]

## 2. 고객DEMO

In [9]:
# 고객DEMO
demo = pd.read_csv('../../data/고객DEMO.txt', encoding='cp949')
demo.head()

고객번호 성별    연령대   거주지역
0     1  M  60세이상   60.0
1     2  M  60세이상  100.0
2     3  M  60세이상   33.0
3     4  F  60세이상   16.0
4     5  M  60세이상  100.0

- demo.고객번호.unique()
> len : 19383
> 1~19383
- demo.성별.unique()
> ['M', 'F']
- demo.연령대.unique()
> `['19세이하', '20세~24세', '25세~29세', '30세~34세', '35세~39세',
'40세~44세', '45세~49세', '50세~54세', '55세~59세', '60세이상']`
- demo.거주지역.unique())
> len : 42\
> `[10.0, 13.0, 16.0, 20.0, 24.0, 27.0, 30.0, 33.0, 36.0, 39.0,
 43.0, 45.0, 47.0, 49.0, 52.0, 55.0, 60.0, 65.0, 69.0, 72.0,
 75.0, 79.0, 82.0, 85.0, 87.0, 100.0, 210.0, 240.0, 270.0, 300.0,
 310.0, 340.0, 360.0, 410.0, 440.0, 460.0, nan, 500.0, 540.0, 570.0,
 610.0, 630.0]`

## 3. 구매상품TR

In [13]:
# 구매상품TR
purprd = pd.read_csv('../../data/구매상품TR.txt', encoding='cp949')
purprd.head()

제휴사    영수증번호  대분류코드  중분류코드    소분류코드   고객번호  점포코드      구매일자  구매시간  구매금액
0   B  8664000     15   1504  B150401  17218    44  20140222    20  2420
1   B  8664000     16   1601  B160101  17218    44  20140222    20  1070
2   B  8664000     16   1602  B160201  17218    44  20140222    20  8060
3   B  8664000     18   1803  B180301  17218    44  20140222    20  6000
4   B  8664001      5    509  B050901  17674    44  20140222    22  1120

## 4. 멤버십여부

In [14]:
# 멤버심 여부
member = pd.read_csv('../../data/멤버십여부.txt', encoding='cp949')
member.head()

고객번호  멤버십명    가입년월
0    11  하이마트  201512
1    21  하이마트  201506
2    37  하이마트  201306
3    43  하이마트  201403
4    44  하이마트  201411

- member.고객번호.unique()
> len: 6498\
> 11 ~ 19383
- member.멤버십명.unique()
> `['다둥이', '더영', '롭스', '하이마트']`
- member.가입년월.unique()
> `[201210, 201211, 201212,
 201301, 201302, 201303, 201304, 201305, 201306, 201307, 201308, 201309, 201310, 201311, 201312,
 201401, 201402, 201403, 201404, 201405, 201406, 201407, 201408, 201409, 201410, 201411, 201412,
 201501, 201502, 201503, 201504, 201505, 201506, 201507, 201508, 201509, 201510, 201511, 201512]`
- member.멤버십명.unique()
> `['다둥이', '더영', '롭스', '하이마트']`

## 5. 상품분류

In [16]:
# 상품분류
prodcl = pd.read_csv('../../data/상품분류.txt', encoding='cp949')
prodcl.head()

제휴사  대분류코드  중분류코드    소분류코드  중분류명  소분류명
0   A      1    101  A010101  일용잡화  위생세제
1   A      1    101  A010102  일용잡화   휴지류
2   A      1    101  A010103  일용잡화  뷰티상품
3   A      1    101  A010104  일용잡화  일용잡화
4   A      1    101  A010105  일용잡화  세트상품

- prodcl.제휴사.unique()
> ['A', 'B', 'C', 'D']
- prodcl.대분류코드.unique()
    - A사: 9
    - B사: 92
    - C사: 17
    - D사: 8
- prodcl.중분류코드.unique()
    - A사: 50
    - B사: 626
    - C사: 98
    - D사: 33
- prodcl.소분류코드.unique()
    - A사: 630
    - B사: 2624
    - C사: 987
    - D사: 154
- prodcl.중분류명.unique()
    - A사: 49
    - B사: 579
    - C사: 98
    - D사: 33
- prodcl.소분류명.unique()
    - A사: 617
    - B사: 2294
    - C사: 987
    - D사: 145

## 6. 채널이용

In [19]:
# 채널이용
channel = pd.read_csv('../../data/채널이용.txt', encoding='cp949')
channel.head()

고객번호           제휴사  이용횟수
0     7  A_MOBILE/APP     4
1    14  A_MOBILE/APP     1
2    42  B_MOBILE/APP    23
3    74  A_MOBILE/APP     1
4    74  B_MOBILE/APP    30

- channel.고객번호.unique()
> len: 6766\
> 7~19382
- channel.제휴사.unique()
> `['A_MOBILE/APP', 'B_MOBILE/APP', 'B_ONLINEMALL', 'C_MOBILE/APP', 'C_ONLINEMALL', 'D_MOBILE/APP']`
- channel.이용횟수.unique()


# 전처리

## 고객DEMO

### 1. 거주지역 코드 -> 지역명

#### 거주지역 코드 - 지역명 맵핑
- 거주지역 코드는 신우편번호 앞 세번자릿수이다.

In [21]:
area_code = {10.: '서울특별시 강북구',
             13.: '서울특별시 도봉구',
             16.: '서울특별시 노원구',
             20.: '서울특별시 중랑구',
             24.: '서울특별시 동대문구',
             27.: '서울특별시 성북구',
             30.: '서울특별시 종로구',
             33.: '서울특별시 은평구',
             36.: '서울특별시 서대문구',
             39.: '서울특별시 마포구',
             43.: '서울특별시 용산구',
             45.: '서울특별시 중구',
             47.: '서울특별시 성동구',
             49.: '서울특별시 광진구',
             52.: '서울특별시 강동구',
             55.: '서울특별시 송파구',
             60.: '서울특별시 강남구',
             65.: '서울특별시 서초구',
             69.: '서울특별시 동작구',
             72.: '서울특별시 영등포구',
             75.: '서울특별시 강서구',
             79.: '서울특별시 양천구',
             82.: '서울특별시 구로구',
             85.: '서울특별시 금천구',
             87.: '서울특별시 관악구',
             100.: '경기도',
             210.: '인천광역시',
             240.: '강원도',
             270.: '충청북도',
             300.: '세종특별자치시',
             310.: '충청남도',
             340.: '대전광역시',
             360.: '경상북도',
             410.: '대구광역시',
             440.: '울산광역시',
             460.: '부산광역시',
             500.: '경상남도',
             540.: '전라북도',
             570.: '전라남도',
             610.: '광주광역시',
             630.: '제주특별자치도'
             }

#### replace 거주지역코드 -> 지역명

In [22]:
# demo_area = demo['거주지역'].map(lambda x:area_code[float(x)])
demo = demo.replace({'거주지역': area_code})
demo

고객번호 성별      연령대       거주지역
0          1  M    60세이상  서울특별시 강남구
1          2  M    60세이상        경기도
2          3  M    60세이상  서울특별시 은평구
3          4  F    60세이상  서울특별시 노원구
4          5  M    60세이상        경기도
...      ... ..      ...        ...
19378  19379  F  20세~24세  서울특별시 동작구
19379  19380  F  20세~24세  서울특별시 마포구
19380  19381  F  40세~44세      부산광역시
19381  19382  F  45세~49세      부산광역시
19382  19383  F  25세~29세       경상남도

[19383 rows x 4 columns]

### 2. 거주지역 nan 값 처리

In [28]:
porc_loc_nan = pd.merge(demo[['고객번호', '거주지역']], purprd[['고객번호', '점포코드']], on='고객번호')

for i in porc_loc_nan[porc_loc_nan.거주지역.isnull()]['고객번호'].unique():
    cust_mode_store = porc_loc_nan[porc_loc_nan.고객번호 == i].점포코드.value_counts().index[0]
    demo.loc[(demo['고객번호'] == i), '거주지역'] = porc_loc_nan[porc_loc_nan.점포코드 == cust_mode_store].거주지역.value_counts().index[0]
display(demo.isnull().sum())
demo

고객번호    0
성별      0
연령대     0
거주지역    0
dtype: int64

고객번호 성별      연령대       거주지역
0          1  M    60세이상  서울특별시 강남구
1          2  M    60세이상        경기도
2          3  M    60세이상  서울특별시 은평구
3          4  F    60세이상  서울특별시 노원구
4          5  M    60세이상        경기도
...      ... ..      ...        ...
19378  19379  F  20세~24세  서울특별시 동작구
19379  19380  F  20세~24세  서울특별시 마포구
19380  19381  F  40세~44세      부산광역시
19381  19382  F  45세~49세      부산광역시
19382  19383  F  25세~29세       경상남도

[19383 rows x 4 columns]

In [31]:
# demo.to_pickle("../../DATA_preprocessed/pp_demo.plk")
pp_demo = pd.read_pickle("../../DATA_preprocessed/pp_demo.plk")
pp_demo

고객번호 성별      연령대       거주지역
0          1  M    60세이상  서울특별시 강남구
1          2  M    60세이상        경기도
2          3  M    60세이상  서울특별시 은평구
3          4  F    60세이상  서울특별시 노원구
4          5  M    60세이상        경기도
...      ... ..      ...        ...
19378  19379  F  20세~24세  서울특별시 동작구
19379  19380  F  20세~24세  서울특별시 마포구
19380  19381  F  40세~44세      부산광역시
19381  19382  F  45세~49세      부산광역시
19382  19383  F  25세~29세       경상남도

[19383 rows x 4 columns]